### ISSUE TITLE
False positives in v0.4.4

### ISSUE THREAD

**Issue by eddie-dunn** [ArnaudBarre/eslint-plugin-react-refresh#30]
I'm getting the following error `Fast refresh only works when a file only exports components. Move your component(s) to a separate file` in a lot of `.tsx` files which only have one export.

Example code for triggering it:

```typescript
const CONSTANT = 1

function Example(props: { prop1: any }) {
  return <div>Example: {CONSTANT}</div>
}

export const ExampleNode3 = {
  type: "node",
  key: "example-node3",
  contents: (props: any) => <Example prop1={props} />,
}

```

Result
```
$ eslint Example.tsx

  1:7   warning  Fast refresh only works when a file only exports components. Move your component(s) to a separate file  react-refresh/only-export-components
  3:10  warning  Fast refresh only works when a file only exports components. Move your component(s) to a separate file  react-refresh/only-export-components

```

Expected result:

1. No warning for `CONSTANT`.
2. ~No warning for `Example`.~ Only a warning for `Example`

**Comment by ArnaudBarre** [comment]
This is a valid warning. Because of the object export the `Example` component cannot be fast refresh and the update will propagate to all importers. This should be detected by the build tool. What tool are you using to bundle React in dev?

**Comment by eddie-dunn** [comment]
> This is a valid warning. Because of the object export the `Example` component cannot be fast refresh and the update will propagate to all importers. 

Shouldn't the warning be on `ExampleNode3` or perhaps _only_ `Example` in that case? 

`CONSTANT` should not get a warning, and the warning seems to disappear if I add a `_` to the variable name.

> This should be detected by the build tool. What tool are you using to bundle React in dev?

Vite.

**Comment by ArnaudBarre** [comment]
Yeah true that `CONSTANT` should not get a warning, I need to improve this.
I think it's better to have the warning on `Example` than `ExampleNode3`.

**Comment by kenlyon** [comment]
I just discovered your plugin and am also trying to understand how it works. The part that confuses me is that the error is not on the thing being exported. If I had `export const Foo = {} // Some big crazy thing.` then I'd understand why it's got an error on it - it shouldn't be exported. But in the above example the `Example` function is not exported so why is it problematic?

I'm using a pattern copied from Material UI (MUI) where there are often some `const` variables declared in the `.tsx` file for things like class names and a `styled()` version of a component. If seems that those are considered problematic, even when not directly exported.

Is the problem more when a `.tsx` _defines_ things other than components? Or is it because they're indirectly exported within the component?

**Comment by ArnaudBarre** [comment]
The rule is "A file that contain at least one react component should not export something else than React components".

Why? Because to Fast Refresh components, the JS modules (i.e. the file) should be re-executed in isolation, and if something is exported, the importers need to get the latest reference without be-ing re-executed. Fast Refresh take cares of 'providing the latest reference' for components exports, but not other exports.

Some framework like Remix can inject more code in to provide this for specific exports, but this cannot be done reliably for random objects.

**Comment by eddie-dunn** [comment]
Thanks for the explanation, @ArnaudBarre . I would suggest you put it in the README, to make it more n00b friendly :)

So, if I understand this correctly, the problem is if a module

1. contains a 'component' (defined by some heuristic such as a function returning JSX, I assume?) 
2. exports a non-component

Is this correct?

**Comment by ArnaudBarre** [comment]
Yeah that's it. Both are based on heuristic which were improved on 0.4.4.
The heuristic for 1 is components is mostly functions returning with starting with an uppercase or result of a memo/forwardRef call.
The heuristic for 2 is name startsWith lower case or is clearly a not a component (object, array)

**Comment by tsareg** [comment]
Hey @ArnaudBarre , any updates on the `CONSTANT` thing? I'd like to use the plugin on a pretty big codebase but it generates tons of errors for the cases like
```
const CONSTANT = 1
```

**Comment by ArnaudBarre** [comment]
Can you provide an example of file where this triggers? Here the warning is not totally wrong, as soon as you have a component inside a file that export something else you need to change something

**Comment by tsareg** [comment]
@ArnaudBarre , looks like the error is a bit different, but basically here is situation:

This will cause warning, eslint points to first line:

```js
const MyConstant = 1234;

export default function doSomething() {
    console.log(MyConstant);
}
```

but this one is fine (though shouldn't be, because file is jsx, but it's not a component)
```js
const myConstant = 1234;

export default function doSomething() {
    console.log(myConstant);
}
```

this one is also fine:
```js
const MY_CONSTANT = 1234;

export default function doSomething() {
    console.log(MY_CONSTANT);
}
```

**Comment by ArnaudBarre** [comment]
Hum I confused can you provide me an example of a JSX file that actually contains JSX and create a warning you think is not correct?
Some rules that the plugin expect because this is common in React:
- `.jsx` should be used when and only when having JSX in the file.
- A Function that return JSX should be in PascalCase

**Comment by tsareg** [comment]
I think I got it:

This will work fine:
```jsx
const MyConstant = 1234;

export default function MyComponent() {
    return <div>{MyConstant}</div>;
}
```

changing function name to lower case will cause warning but that's **expected behavior**:
```jsx
const MyConstant = 1234;

export default function myComponent() {
    return <div>{MyConstant}</div>;
}
```
However the misleading part is that eslint complains about first line.

Now
```jsx
const myConstant = 1234;

export default function myComponent() {
    return <div>{myConstant}</div>;
}
```
Will not cause any warnings. Is that expected?

**Comment by ArnaudBarre** [comment]
I've accumulated multiple issues like this where there is cases where the warning is not optimal.
The problem is that when the file export both a component and a non component, it's hard to say it the one that should be move out if the non component (shared utils) or the component (route configuration entry point).

I took some time to rethink it, but again I didn't find any meaningful changes that makes it worth breaking existing ignore comments.

@tsareg Note that the rule don't check that function that return JSX actually follow the React requirement to be PascalCase, so the last case is expected to not get any warning from this rule. This could be something enforce by another rule.
